File: algo/problems/0049-group-anagrams/scala/GroupAnagrams.scala
object GroupAnagrams:
  /** 用 groupBy(排序后的字符串) 一行聚合（Scala 最简写法）。O(n·k·log k)。 */
  def groupAnagrams(strs: Array[String]): List[List[String]] =
    strs.groupBy(s => s.sorted).values.map(_.toList).toList

File: algo/problems/0049-group-anagrams/go/groupanagrams.go
package groupanagrams

import "sort"

// GroupAnagrams 以"排序后的字符串"为键聚合。O(n·k·log k)。
func GroupAnagrams(strs []string) [][]string {
	groups := make(map[string][]string)
	for _, s := range strs {
		b := []byte(s)
		sort.Slice(b, func(i, j int) bool { return b[i] < b[j] })
		key := string(b)
		groups[key] = append(groups[key], s)
	}
	res := make([][]string, 0, len(groups))
	for _, g := range groups {
		res = append(res, g)
	}
	return res
}
